File: ios/Packages/TallyFeatureAPIClient/Sources/TallyFeatureAPIClient/EntryDefaults.swift
import Foundation

/// Utilities for calculating smart initial values for new entries.
/// Uses the last 2 weeks of data to suggest starting values.
public enum EntryDefaults {
    
    /// Calculate initial value for a new entry based on recent history.
    ///
    /// - For simple count: returns the average count from the last 14 days, rounded to nearest 5.
    /// - For sets: returns the average of first-set values from the last 14 days, rounded.
    ///
    /// - Parameters:
    ///   - entries: All entries for the challenge
    ///   - countType: The challenge's count type
    /// - Returns: The suggested initial value (minimum 1)
    public static func calculateInitialValue(
        entries: [Entry],
        countType: CountType = .simple
    ) -> Int {
        let twoWeeksAgo = Calendar.current.date(byAdding: .day, value: -14, to: Date())!
        let cutoffDate = ISO8601DateFormatter.string(from: twoWeeksAgo, timeZone: .current, formatOptions: .withFullDate)
        
        let recentEntries = entries.filter { $0.date >= cutoffDate }
        
        guard !recentEntries.isEmpty else {
            return 1
        }
        
        if countType == .sets {
            // For sets mode: average of first-set values
            let firstSetValues = recentEntries
                .compactMap { $0.sets?.first }
            
            if firstSetValues.isEmpty {
                return calculateSimpleAverage(entries: recentEntries)
            }
            
            let avg = Double(firstSetValues.reduce(0, +)) / Double(firstSetValues.count)
            return max(1, Int(avg.rounded()))
        }
        
        return calculateSimpleAverage(entries: recentEntries)
    }
    
    /// Calculate simple average, rounded to nearest 5.
    private static func calculateSimpleAverage(entries: [Entry]) -> Int {
        guard !entries.isEmpty else { return 1 }
        
        let sum = entries.reduce(0) { $0 + $1.count }
        let avg = Double(sum) / Double(entries.count)
        
        // Round to nearest 5 for cleaner numbers
        let rounded = Int((avg / 5.0).rounded() * 5)
        return max(1, rounded != 0 ? rounded : Int(avg.rounded()))
    }
    
    /// Get the last set value for the "add another set" flow.
    ///
    /// - Parameter currentSets: The current sets array in the form
    /// - Returns: The last set value, or 1 if empty/invalid
    public static func getLastSetValue(currentSets: [Int]) -> Int {
        guard let last = currentSets.last, last > 0 else {
            return 1
        }
        return last
    }
    
    /// Calculate sets statistics for dashboard display.
    ///
    /// - Parameter entries: All entries (may include entries without sets)
    /// - Returns: Tuple with bestSet and avgSetValue, or nil values if no sets data
    public static func calculateSetsStats(
        entries: [Entry]
    ) -> (bestSet: (value: Int, date: String, entryId: String)?, avgSetValue: Double?) {
        // Collect all individual set values with metadata
        var allSets: [(value: Int, date: String, entryId: String)] = []
        
        for entry in entries {
            if let sets = entry.sets, !sets.isEmpty {
                for setVal in sets {
                    allSets.append((value: setVal, date: entry.date, entryId: entry.id))
                }
            }
        }
        
        guard !allSets.isEmpty else {
            return (nil, nil)
        }
        
        // Find best set
        let bestSet = allSets.max(by: { $0.value < $1.value })
        
        // Calculate average
        let sum = allSets.reduce(0) { $0 + $1.value }
        let avgSetValue = (Double(sum) / Double(allSets.count) * 10).rounded() / 10
        
        return (bestSet, avgSetValue)
    }
}

private extension ISO8601DateFormatter {
    static func string(from date: Date, timeZone: TimeZone, formatOptions: ISO8601DateFormatter.Options) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = formatOptions
        formatter.timeZone = timeZone
        return formatter.string(from: date)
    }
}

File: tally-android/core/network/src/main/java/com/tally/core/network/EntryDefaults.kt
package com.tally.core.network

import java.time.LocalDate
import kotlin.math.roundToInt

/**
 * Utilities for calculating smart initial values for new entries.
 * Uses the last 2 weeks of data to suggest starting values.
 */
object EntryDefaults {
    
    /**
     * Calculate initial value for a new entry based on recent history.
     *
     * - For simple count: returns the average count from the last 14 days, rounded to nearest 5.
     * - For sets: returns the average of first-set values from the last 14 days, rounded.
     *
     * @param entries All entries for the challenge
     * @param countType The challenge's count type
     * @return The suggested initial value (minimum 1)
     */
    fun calculateInitialValue(
        entries: List<Entry>,
        countType: CountType = CountType.SIMPLE
    ): Int {
        val twoWeeksAgo = LocalDate.now().minusDays(14)
        val cutoffDate = twoWeeksAgo.toString() // ISO format YYYY-MM-DD
        
        val recentEntries = entries.filter { it.date >= cutoffDate }
        
        if (recentEntries.isEmpty()) {
            return 1
        }
        
        return when (countType) {
            CountType.SETS -> {
                // For sets mode: average of first-set values
                val firstSetValues = recentEntries
                    .mapNotNull { it.sets?.firstOrNull() }
                
                if (firstSetValues.isEmpty()) {
                    calculateSimpleAverage(recentEntries)
                } else {
                    val avg = firstSetValues.average()
                    maxOf(1, avg.roundToInt())
                }
            }
            else -> calculateSimpleAverage(recentEntries)
        }
    }
    
    /**
     * Calculate simple average, rounded to nearest 5.
     */
    private fun calculateSimpleAverage(entries: List<Entry>): Int {
        if (entries.isEmpty()) return 1
        
        val avg = entries.map { it.count }.average()
        
        // Round to nearest 5 for cleaner numbers
        val rounded = ((avg / 5.0).roundToInt() * 5)
        return maxOf(1, if (rounded != 0) rounded else avg.roundToInt())
    }
    
    /**
     * Get the last set value for the "add another set" flow.
     *
     * @param currentSets The current sets array in the form
     * @return The last set value, or 1 if empty/invalid
     */
    fun getLastSetValue(currentSets: List<Int>): Int {
        val last = currentSets.lastOrNull()
        return if (last != null && last > 0) last else 1
    }
    
    /**
     * Calculate sets statistics for dashboard display.
     *
     * @param entries All entries (may include entries without sets)
     * @return Pair of bestSet and avgSetValue, or null values if no sets data
     */
    fun calculateSetsStats(entries: List<Entry>): Pair<Triple<Int, String, String>?, Double?> {
        // Collect all individual set values with metadata
        data class SetData(val value: Int, val date: String, val entryId: String)
        
        val allSets = mutableListOf<SetData>()
        
        for (entry in entries) {
            entry.sets?.forEach { setVal ->
                allSets.add(SetData(setVal, entry.date, entry.id))
            }
        }
        
        if (allSets.isEmpty()) {
            return Pair(null, null)
        }
        
        // Find best set
        val bestSet = allSets.maxByOrNull { it.value }?.let {
            Triple(it.value, it.date, it.entryId)
        }
        
        // Calculate average (one decimal place)
        val avgSetValue = (allSets.map { it.value }.average() * 10).roundToInt() / 10.0
        
        return Pair(bestSet, avgSetValue)
    }
}
